File: src/devtwin/adapters/php.py
"""PHP ecosystem adapter: Composer-based projects."""

from __future__ import annotations

import json
from pathlib import Path

from devtwin.adapters.base import EcosystemAdapter, extract_version
from devtwin.core.models import DependencyInfo, HealthIssue, Presence, RuntimeInfo, Severity
from devtwin.core.runner import run_command, which


def _read_composer_json(root: Path) -> dict:
    f = root / "composer.json"
    if not f.exists():
        return {}
    try:
        return json.loads(f.read_text(encoding="utf-8", errors="ignore"))
    except (OSError, json.JSONDecodeError):
        return {}


class PhpAdapter(EcosystemAdapter):
    ecosystem = "php"

    def detect(self, root: Path) -> bool:
        return (root / "composer.json").exists()

    def _required_version(self, root: Path) -> str | None:
        require = _read_composer_json(root).get("require", {})
        return require.get("php") if isinstance(require, dict) else None

    def inspect_runtime(self, root: Path) -> list[RuntimeInfo]:
        required = self._required_version(root)
        php_path = which("php")
        if php_path is None:
            return [
                RuntimeInfo(
                    name="php",
                    presence=Presence.NOT_INSTALLED,
                    required_version=required,
                    source="composer.json [require.php]" if required else None,
                )
            ]
        result = run_command([php_path, "--version"], timeout=5)
        installed = extract_version(result.stdout) if result.available else None
        return [
            RuntimeInfo(
                name="php",
                presence=Presence.INSTALLED if installed else Presence.UNKNOWN,
                installed_version=installed,
                required_version=required,
                path=php_path,
                source="composer.json [require.php]" if required else None,
            )
        ]

    def inspect_dependencies(self, root: Path) -> DependencyInfo | None:
        if not (root / "composer.json").exists():
            return None
        lockfile = root / "composer.lock"
        return DependencyInfo(
            ecosystem="php",
            manager="composer",
            lockfile="composer.lock" if lockfile.exists() else None,
            lockfile_present=lockfile.exists(),
            manifest_present=True,
        )

    def inspect_tests(self, root: Path) -> list[str]:
        scripts = _read_composer_json(root).get("scripts", {})
        if isinstance(scripts, dict) and "test" in scripts:
            return ["composer test"]
        if (root / "phpunit.xml").exists() or (root / "phpunit.xml.dist").exists():
            return ["vendor/bin/phpunit"]
        return []

    def health_checks(self, root: Path, runtimes: list[RuntimeInfo]) -> list[HealthIssue]:
        issues = []
        for rt in runtimes:
            if rt.name == "php" and rt.presence == Presence.NOT_INSTALLED:
                issues.append(
                    HealthIssue(
                        severity=Severity.CRITICAL,
                        code="php.not_installed",
                        title="PHP not found",
                        message="No `php` executable was found on PATH.",
                        evidence=["which php -> not found"],
                        recommendation="Install PHP via https://www.php.net/manual/en/install.php.",
                    )
                )
        return issues

File: mcphub/devtwin/src/adapters/php.ts
/** PHP ecosystem adapter: Composer-based projects. */

import { join } from 'node:path';

import {
  Presence,
  Severity,
  type DependencyInfo,
  type HealthIssue,
  type RuntimeInfo,
} from '../core/models.js';
import { runCommand, which } from '../core/runner.js';
import { pathExists, readJsonFile } from '../system/filesystem.js';
import { EcosystemAdapter, extractVersion } from './base.js';

/** Read composer.json, degrading to `{}` on a missing file or malformed JSON. */
function readComposerJson(root: string): Record<string, unknown> {
  const file = join(root, 'composer.json');
  if (!pathExists(file)) {
    return {};
  }
  const data = readJsonFile<Record<string, unknown>>(file);
  if (data === null || typeof data !== 'object' || Array.isArray(data)) {
    return {};
  }
  return data;
}

export class PhpAdapter extends EcosystemAdapter {
  readonly ecosystem = 'php';

  detect(root: string): boolean {
    return pathExists(join(root, 'composer.json'));
  }

  private requiredVersion(root: string): string | null {
    const require = readComposerJson(root)['require'];
    // Only a JSON object counts -- a scalar or array `require` is ignored.
    if (typeof require !== 'object' || require === null || Array.isArray(require)) {
      return null;
    }
    const php = (require as Record<string, unknown>)['php'];
    return typeof php === 'string' ? php : null;
  }

  async inspectRuntime(root: string): Promise<RuntimeInfo[]> {
    const required = this.requiredVersion(root);
    const phpPath = which('php');
    if (phpPath === null) {
      return [
        {
          name: 'php',
          presence: Presence.NOT_INSTALLED,
          installed_version: null,
          required_version: required,
          path: null,
          source: required ? 'composer.json [require.php]' : null,
        },
      ];
    }
    const result = await runCommand([phpPath, '--version'], { timeout: 5 });
    const installed = result.available ? extractVersion(result.stdout) : null;
    return [
      {
        name: 'php',
        presence: installed ? Presence.INSTALLED : Presence.UNKNOWN,
        installed_version: installed,
        required_version: required,
        path: phpPath,
        source: required ? 'composer.json [require.php]' : null,
      },
    ];
  }

  async inspectDependencies(root: string): Promise<DependencyInfo | null> {
    if (!pathExists(join(root, 'composer.json'))) {
      return null;
    }
    const lockfilePresent = pathExists(join(root, 'composer.lock'));
    return {
      ecosystem: 'php',
      manager: 'composer',
      lockfile: lockfilePresent ? 'composer.lock' : null,
      lockfile_present: lockfilePresent,
      manifest_present: true,
      installed: Presence.UNKNOWN,
      dependency_count: null,
      dev_dependency_count: null,
      notes: [],
    };
  }

  async inspectTests(root: string): Promise<string[]> {
    const scripts = readComposerJson(root)['scripts'];
    if (
      typeof scripts === 'object' &&
      scripts !== null &&
      !Array.isArray(scripts) &&
      'test' in (scripts as Record<string, unknown>)
    ) {
      return ['composer test'];
    }
    if (pathExists(join(root, 'phpunit.xml')) || pathExists(join(root, 'phpunit.xml.dist'))) {
      return ['vendor/bin/phpunit'];
    }
    return [];
  }

  async healthChecks(_root: string, runtimes: RuntimeInfo[]): Promise<HealthIssue[]> {
    const issues: HealthIssue[] = [];
    for (const rt of runtimes) {
      if (rt.name === 'php' && rt.presence === Presence.NOT_INSTALLED) {
        issues.push({
          severity: Severity.CRITICAL,
          code: 'php.not_installed',
          title: 'PHP not found',
          message: 'No `php` executable was found on PATH.',
          evidence: ['which php -> not found'],
          recommendation: 'Install PHP via https://www.php.net/manual/en/install.php.',
          confidence: null,
        });
      }
    }
    return issues;
  }
}
